### DISCUSSION TITLE
Move UR5e arm with MoveIt Servo on ROS2 Humble

### DISCUSSION THREAD

**Discussion by fornaroale** [UniversalRobots/Universal_Robots_ROS2_Driver#827] [Q&A]
Hello,
I'm having issues trying to setup my "digital" UR5e robot (RViz and MoveIt Humble, no real robot) on ROS2. I currently followed the following steps (given in the documentation):
- `ros2 launch ur_robot_driver ur_control.launch.py ur_type:=ur5e robot_ip:=yyy.yyy.yyy.yyy use_fake_hardware:=true launch_rviz:=false`
- `ros2 launch ur_moveit_config ur_moveit.launch.py ur_type:=ur5e launch_rviz:=true`

I then switched the controller to forward_position_controller and called the servo_node service, as in ROS2 Humble documentation:
- `ros2 control switch_controllers --deactivate scaled_joint_trajectory_controller --activate forward_position_controller`
- `ros2 service call /servo_node/start_servo std_srvs/srv/Trigger`

.. and after these, I'm publishing the following data, as advised in one of the issues of this repository:
`ros2 topic pub /servo_node/delta_twist_cmds geometry_msgs/msg/TwistStamped "{ header: { stamp: 'now' },  twist: {linear: {x: -0.1}, angular: {  }}}" -r 10`

Unfortunately, the robot is not moving in RViz.

I'd like to teleoperate the (virtual) robot and, as far as I understood, this controller is what I need (what I need to realise in the future is an external python script which will publish cartesian movements on this topic, making the arm moving in RViz). I'm sorry for the missing label, but I couldn't find an adequate one.

Could I kindly ask you to help me with this problem? Thank you.

**Comment by fmauch** [comment]
Not being able to test this at the time of writing (I am on the go without having moveit_servo available): I think you could benefit from passing `use_fake_hardware:=true` to the moveit startup, as well:

```
ros2 launch ur_moveit_config ur_moveit.launch.py ur_type:=ur5e launch_rviz:=true use_fake_hardware:=true
```

This will tell MoveIt! to use the `joint_trajectory_controller` as opposed to the `scaled_joint_trajectory_controller`.

However, making the tests I currently can do it seems that the scaled JTC is actually working using fake hardware, so this might actually be not necessary at all and we could remove this flag from our launchfile...

**Reply by fornaroale** [reply]
Unfortunately, passing `use_fake_hardware:=true` to the moveit startup didn't work out. However, using cmds like the following one:
`ros2 topic pub /forward_position_controller/commands std_msgs/msg/Float64MultiArray "{ data: [ 0.3, 0.8, 0.2, 0, 0, 0 ] }" --once`
 instead of the one in the question, did work out and made the robot move in RViz.

To be honest I don't think I'm following the right path to achieve what I need (real-time control sending target cartesian positions). I've been checking out [this repository](https://github.com/stefanscherzinger/cartesian_controllers_universal_robots) which seems to be more appropriate, but unfortunately it doesn't seem to provide collision checking (I don't need it, but I need to know when the robot gets in joint configuration in which some joints collide with themselves), and I've not been lucky using moveit_servo, because I can't find any tutorial to work with [UniversalRobot drivers for ROS2](https://github.com/UniversalRobots/Universal_Robots_ROS2_Driver/) using real-time cartesian targets (like move 0.1 in x, move 0.2 in y, and so on.., or giving triples like {x:1,y:1,z:0.5}). The keyboard example from ROS2 Humble is somewhat hard-coded with Panda robot.

Thank you for your answer.

**Comment by jozefkosecky** [comment]
@fornaroale 
Please check this: https://github.com/LucaBross/simple_moveit2_universal_robots_movement. It works for me. Also, if you encounter other problems, you can refer to this: https://github.com/UniversalRobots/Universal_Robots_ROS2_Driver/issues/457

**Reply by fornaroale** [reply]
Right now I'm using [these cartesian controllers](https://github.com/fzi-forschungszentrum-informatik/cartesian_controllers/tree/ros2), achieving what I need. The problem is that I need to adjust my target in real-time, and I don't know if MoveIt is adequate for such a task: it's for this reason that I chose the cartesian controller, which works without all the MoveIt stuff and allows me to change the target cartesian point of the end effector basically in real-time, by simply publishing on a topic. Thank you for your answer!

**Reply by jozefkosecky** [reply]
Many years ago, I also changed the target point in real-time. Back then, I used ROS1 with MoveIt to achieve that. Now, I'm working on a similar project where I will change the target point in real time again. I will go with MoveIt, but I will also check that. Can I reach out to you when things go bad?

**Reply by fornaroale** [reply]
I'm new to ROS and have absolutely no expertise in the robotics field. I think I've very little to say, what I could suggest you is to check the issues in the repository; I learned useful things looking there. 
However, that said, If I can be of any help, It will be a pleasure, but I highly doubt! 🥲
